### ISSUE TITLE
[Q] How does `wandb.log` work with JAX's asynchronous dispatch?

### ISSUE THREAD

**Issue by samuela** [wandb/wandb#3690]
Many frameworks like JAX utilize [asynchronous dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html) for improved accelerator utilization. How does this work with `wandb.log`? If I log a metric on every batch does that have the effect of blocking execution until the logged metric's `DeviceArray` becomes available?

There's also the question of buffering network requests. Does every `wandb.log` invoke a separate network request? If they are batched, what is the batching strategy? Are network requests blocking?

**Comment by samuela** [comment]
PyTorch follows a similar lazy execution model: https://pytorch.org/docs/stable/notes/cuda.html#asynchronous-execution

**Comment by samuela** [comment]
TF also has a similar concept: https://jonathan-hui.medium.com/tensorflow-eager-execution-v-s-graph-tf-function-6edaa870b1f1

**Comment by samuela** [comment]
Indeed, I am observing a near 2x slowdown when calling `wandb.log` between every batch. Is this performance gotcha documented anywhere?

**Comment by vanpelt** [comment]
@samuela we do not block on any network calls.  When `wandb.init` is called we spawn a separate process on the same machine you're logging from which queues and batches requests asynchronously streaming them to our server.

When you call `wandb.log` we need to know what values you're logging so if you pass tensors to `wandb.log` we attempt to materialize them which could incur a slowdown as data is copied from the GPU to the CPU.  You should be able to optimize this penalty by being smart about how your caching / updating any metrics.  Most frameworks like PyTorch Lightning do this automatically for you.  With a framework like Jax you'll need to optimize the way you access your metrics manually.

For us to understand the specific slowdown you're experiencing we would need a reproducible script to profile.  Another thing to try would be setting `WANDB_MODE=disabled`.  This makes all calls to W&B api's NOOPs that don't go through our regular pipelines which could tell us if the slowdown is in W&B or it's in the way you're pulling metrics.

The only other overhead incurred by `wandb.log` calls is serializing the data before it's put on the queue.  This should be negligible but could be larger if you're pushing very large histograms or large data structures through `wandb.log`.

**Comment by vanpelt** [comment]
Specifically for JAX [we call](https://github.com/wandb/client/blob/b67be917facb1ccac993a7898c72c00cb8a6aad9/wandb/util.py#L472-L475) `jax.device_get(obj)` if a value in call to `wandb.log` is a JAX tensor.  You can find other examples of [materializing lazy tensors here](https://github.com/wandb/client/blob/b67be917facb1ccac993a7898c72c00cb8a6aad9/wandb/util.py#L575-L599).

**Comment by samuela** [comment]
Thanks for explaining @vanpelt! This def helps me navigate using wandb appropriately. I'll stick to only calling `wandb.log` at the end of epochs for now.
